### ISSUE TITLE
Writes concurrent with a topology change may experience reduced effective CL

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#5252]
The problem is that some acknowledged writes may end up being stored on fewer replicas than the CL of the request requires after a topology change. For example, for RF=3 keyspace a CL=QUORUM write may end up being present on only 1 replica.

Topology changes are currently performed like this:
 1) the topology change coordinator (e.g. the joining node) advertises the change by changing the STATUS application state for one of the nodes (usually itself)
 2) the topology change coordinator waits for `ring_delay` (default: 30 s) so that the gossip protocol propagates the change of state to all the nodes
 3) streaming is done
 4) STATUS is changed again to indicate the end of the process

A topology change may change the replica set for a given key by replacing one of the existing replicas with another one.

nodes = {A, B, C}

The problem arises if we have a coordinator-side write which spans both the coordinator noticing the topology change and the start of streaming:

1. Node D starts to join the cluster, sets STATUS=BOOT
2. CQL insert of key K with CL=QUORUM starts on node A, the write handler uses endpoints A, B, and C.
3. Node A notices that D joins, updates pending_endpoints, D replaces A for key K
4. Streaming starts
5. Streaming goes past K
6. Replicas A and B acknowledge the write of K from step 2. The write doesn't reach C. The CQL insert succeeds.
7. Streaming ends. Now the replicas for K are {B, C, D}, but only B has the write

What makes this scenario rare is that steps 1 and 4 are separated by `ring_delay`, so for the problem to happen either:
 1) gossip propagation to A needs to be significantly delayed (step 3 is close to step 4)
 2) or, the write is executed with a large timeout and its completion is delayed (step 2 is close to step 1, but the distance between step 2 and step 6 is within the timeout).

In addition to the above, the write must not reach one of the replicas at all.

The fix suggested by Gleb is to have the topology change coordinator synchronize with each node before it starts streaming by issuing a call which makes sure the node sees the new topology and waits for all requests started earlier (which may use the old topology). Requests would be tracked using `phased_barrier`. This makes correctness insensitive to the timing of events.

This is also needed to prevent violation of CAS register (LWT) consistency in a similar scenario where `accept` calls to replicas are delayed:

1. node D starts to join, will replace A for key K
2. CAS on key K starts on node A, uses {A, B, C} as replicas
3. `prepare` succeeds on A and B only
4. `accept` starts on A and B but hangs
5. pending endpoints change on A, streaming starts, goes past K
6. streaming of system.paxos goes past K (didn't see any entry for K)
7. `accept` on A and B completes
8. CAS coordinator commits on A and B, the CAS is ACKed
9. range movement completes, {B, C, D} are the new replicas but the mutation of the CAS is applied only on B

The fact that the CAS was committed to a minority is a problem. Now, if another CAS commits a conflicting transaction using {C, D}, we will end up with the effects of two conflicting transactions being visible to reads.

**Comment by slivne** [comment]
@tgrabiec / @gleb-cloudius how does C* handle this - is there an issue for this in C* ?

**Comment by kostja** [comment]
This is a duplicate of #5154. Streaming must not start before the joining node has confirmed that all affected nodes has received its tokens.

**Comment by tgrabiec** [comment]
@kostja Are you saying that #5154 is supposed to fix this issue? In its current form, this issue is out of the scope of that pull request. For example, it doesn't address the problem of waiting for pending requests which are still using the old topology. The main goal of #5154 is to allow fast bootstrap in the happy path without changing current safety guarantees much.

**Comment by kostja** [comment]
1) I checked the latest patch for #5154 and indeed, it simply times out in case of "unhappy" path. I believe there should be no "unhappy" path at all, wait_for_gossip_to_settle() should not return until at least the *token metadata* hashes match. If they don't match for an extended period (300 seconds?) the operation can abort. As long as we add one or two nodes at a time I can't imagine a condition when this would be a usability issue. This would make sure the node doesn't start streaming before other nodes are aware of it, and it is included in quorum/required_participants. If it's outside of scope of #5154, then it must be part of the scope of this bug. In any case, the fix should reside in gossip subsystem, not in LWT.
The reason I pointed out #5154 is that it looks this part of the issue looks like a much bigger fish to fry than this bug.

2)  I have a really hard time seeing how a CAS operation can get stuck while node addition, which includes committing new node tokens, is allowed to proceed. I agree we must consider and address this possibility nevertheless.
I agree with Gleb's solution, and we may require it for other cases, too (e.g., waiting for pending paxos requests before accepting a DDL on the affected table).  It seems we should indeed provide wait_for_started_paxos_rounds() API to streaming.
However, I think the probability, and hence impact, of this event, if 1) is fixed, is extremely, almost negligibly, small, so it is not worth looking at until 1).

**Comment by gleb-cloudius** [comment]
@kostja the problem and the proposed fix is not CAS specific. @vladzcloudius started to implemented it once, but he probably has other things on his plate right now.

**Comment by kostja** [comment]
A discussion clarifying why wait_for_started_paxos_rounds() should wait on all nodes:

@kostja  4:57 PM
#5252 is about two things, ensuring that gossip never falls through to streaming
without acknowledging that new tokens are out thre, and that streaming never
starts before acknowledging that all transactions that started with old tokens
have finished (edited) 
ok, so let's distill the issues besides the two above that are still remaining

@tgrabiec  4:59 PM
if 5252 is fixed by waiting for transactions, we still have a problem caused by
the fact that transactions may timeout on coordinator-side, but not on replica
side.

@kostja  4:59 PM
why is it an issue?

@tgrabiec  5:01 PM
if replica-side transaction commits after streaming goes past that key, we may
end up with minority of replicas seeing the commit and thus violate LWT
invariants, which allows you to violate linearizability

@kostja  5:01 PM
ok, let's slow down a bit and describe the scenario from the beginning.

0. Nodes A, B,C are replicas, C is temporarily partitioned out.
1. New node D advertises its tokens.
2. A new transaction starts on B which is yet unaware of the new tokens.
3. The transaction advances on A and B successfuly to "accepted" phase, C is
  not responding, times out on B in "learn" phase without doing any changes,
and freezes on A in the beginning of "learn" phase.
4. #5252 is fixed, so the new node acknowledges that other nodes got its tokens,
including node C.
5. Streaming starts, goes past the changed key. No node is
repaired since no node has its base table changed yet - the transaction on A
is frozen before having changed the base table.
6. A learns/commits the transaction.
7. Due to a topology change the replica that committed the transaction is a
minority: B is replaced by D, and since system.paxos is not streamed,
D's system.paxos is empty.
8. A is temporarily partitioned out, C and D go on and accept a change that
is not based on an existing transaction which has committed on A. E.g. a condition
which is supposed to evaluate to true because of changes on A, evaluates to false.
We have a linearizable history split. 

@tgrabiec  5:22 PM
Because we don't repair system.paxos during streaming, the transaction may end
up in "accepted" state in majority of old participants, but in minority of new
participants. So you can execute later transactions using majority of new
participants which didn't see the old transaction.

@kostja  5:33 PM
alright, so it still seems tobe #5252, not #5251 then
wait_for_started_paxos_rounds() which I suggested to add in #5252, should wait
for pending transactions on all nodes, not just coordinator nodes, before
starting streaming. 

@tgrabiec  5:34 PM
If you mean that it waits for replica-side transactions as well, then yes 5251
will be fixed as well.

@tgrabiec  5:51 PM
There is one problem with this approach, all operations would have to change
semantics to failure means not-applied rather than "don't know". There is one
known place which doesn't respect this. A write may be applied to the commit
log but fail to apply to memtable (and fail altogether), and later reappear.
If a write reappears after you let streaming start, we're back to the race and
can end up committing something which was not streamed.
The barrier like wait_for_started_paxos_rounds() would wait for all ops to
either fail or succeed. If failure doesn't mean that the op finished, it's as
if you didn't wait for it.

**Comment by vladzcloudius** [comment]
> @kostja the problem and the proposed fix is not CAS specific. @vladzcloudius started to implemented it once, but he probably has other things on his plate right now.

@gleb-cloudius @kostja I was actually trying to rebase that series on the current master a few days ago but failed miserably due to tectonic changes in the OPTIONS related code.

I have a version that have addressed all comments - I just need to rebase, test and send it out.

The relevant thread is "gossip: gracefully shut the node down"

**Comment by kbr-** [comment]
@kostja @tgrabiec

I think there are other solutions to this problem. We can find them by looking at which invariants of the Synod protocol could be broken today, and how to make sure that we maintain them instead.

Correctness of Synod depends on (among other things) what Lamport calls invariant B2 in his paper, i.e. the quorums of any two ballots must have a non-empty intersection.

The problem here, as Tomek pointed out, and Kostja described with an example, is that we allow two ballots *in the same Synod instance* to have non-intersecting quorums: in one ballot we use {A, B} as a quorum, in the other we use {C, D} as a quorum.

Here's two properties that any solution must preserve:
1. If we are still running a Synod instance that could have used {A, B} as quorum, don't allow {C, D} to be a quorum. The allowed quorums are: {A, B}, {A, C}, {B, C}, {A, B, D}, {A, C, D}, {B, C, D}. (any two of these sets intersect)
2. If we want {C, D} to become a valid quorum, we have to make sure to finish the previous Synod instance and start another one. So no ballot in the history of all ballots uses A as an acceptor (in this new Synod instance).

If I understand correctly, Kostja's proposition is to make sure that the previous Synod instance is finished by calling `wait_for_started_paxos_rounds()` before starting streaming.
But we don't need to be so strict. We can keep running Synod instances that started before bootstrapping started; just *don't consider {C, D} to be a valid quorum* during streaming. In general, during streaming, any valid quorum must contain a majority of old replicas.
However, before any node starts using the new tokens (i.e. changes the pending tokens of the new node to normal tokens), here is the place we absolutely must make sure that Synod instances that started before, have finished (including the learning phase). All new Synod instances can then use the new set of acceptors, and {C, D} will be a valid quorum.

---

Here's another approach at the problem, by looking at it a bit differently.
Every node is running an acceptor. The node is not the acceptor, it is just where the acceptor "resides". So, for example, node A has acceptor A', node B has acceptor B', and node C has acceptor C'.
When we boostrap node D, and change the replica set from {A, B, C} to {B, C, D}, instead of creating a new acceptor D' on D, we migrate the acceptor from A: i.e. D will have acceptor A', and A will have no acceptor.
In terms of implementation, we would move relevant data from system.paxos on A to system.paxos on D.
To implement this we would need to be very careful so that A' doesn't run concurrently on both A and D. So it would have to proceed in steps: we first block A from modifying A's state, then we migrate A' to D, then we allow D to modify A' state.
This way, Synod invariants are preserved trivially: the set of acceptors was {A', B', C'}, and after bootstrapping it didn't change. So the allowed quorums are still the same: {A', B'}, {A', C'}, and {B', C'}. It's just that the acceptor A' is now running on a different node.

**Comment by denesb** [comment]
@kbr- what is this "Synod protocol"? Google turning up nothing useful for it. I don't recall hearing of it before, is this something we are planning on using?

**Comment by kbr-** [comment]
@denesb 

It's also called "single-decree Paxos". It's what we are running (more like a variant of it).
https://lamport.azurewebsites.net/pubs/lamport-paxos.pdf page 4. I highly recommend reading the whole paper, it's very enjoyable (this is the original Paxos paper).

**Comment by denesb** [comment]
Thanks @kbr- will do. I read the "Paxos made simple" paper, which has no mention of this Synod as I recall.

**Comment by kbr-** [comment]
Yeah, the terminology today is a bit mixed.
The protocol in section 2 of "Paxos made simple" was originally called Synod, and the state machine replication algorithm in section 3 was the actual Paxos. This "multi-decree Paxos" is described in much more detail in the original paper.

**Comment by slivne** [comment]
relevant after asias's patch for quick bootstrap

**Comment by tgrabiec** [comment]
> @kostja @tgrabiec
> 
> I think there are other solutions to this problem. We can find them by looking at which invariants of the Synod protocol could be broken today, and how to make sure that we maintain them instead.
> 
> Correctness of Synod depends on (among other things) what Lamport calls invariant B2 in his paper, i.e. the quorums of any two ballots must have a non-empty intersection.
> 
> The problem here, as Tomek pointed out, and Kostja described with an example, is that we allow two ballots _in the same Synod instance_ to have non-intersecting quorums: in one ballot we use {A, B} as a quorum, in the other we use {C, D} as a quorum.
> 
> Here's two properties that any solution must preserve:
> 
> 1. If we are still running a Synod instance that could have used {A, B} as quorum, don't allow {C, D} to be a quorum. The allowed quorums are: {A, B}, {A, C}, {B, C}, {A, B, D}, {A, C, D}, {B, C, D}. (any two of these sets intersect)
> 2. If we want {C, D} to become a valid quorum, we have to make sure to finish the previous Synod instance and start another one. So no ballot in the history of all ballots uses A as an acceptor (in this new Synod instance).
> 
> If I understand correctly, Kostja's proposition is to make sure that the previous Synod instance is finished by calling `wait_for_started_paxos_rounds()` before starting streaming.

hmm, I don't think it's an accurate description of Kostja's proposal. We're not stopping the Synod and start a new one, while we wait we still allow new transactions using the new set of quorums to start and interact with results of transactions using the old set of quorums. The waiting is only a barrier which tells us when transactions using the old set of quorums are no longer (and will not be) running.

Both new and old sets of quorums overlap as you say in point 1.


> But we don't need to be so strict. We can keep running Synod instances that started before bootstrapping started; just _don't consider {C, D} to be a valid quorum_ during streaming. In general, during streaming, any valid quorum must contain a majority of old replicas.
> However, before any node starts using the new tokens (i.e. changes the pending tokens of the new node to normal tokens), here is the place we absolutely must make sure that Synod instances that started before, have finished (including the learning phase). All new Synod instances can then use the new set of acceptors, and {C, D} will be a valid quorum.

What do you mean by stopping synod instances here?

> 
> Here's another approach at the problem, by looking at it a bit differently.
> Every node is running an acceptor. The node is not the acceptor, it is just where the acceptor "resides". So, for example, node A has acceptor A', node B has acceptor B', and node C has acceptor C'.
> When we boostrap node D, and change the replica set from {A, B, C} to {B, C, D}, instead of creating a new acceptor D' on D, we migrate the acceptor from A: i.e. D will have acceptor A', and A will have no acceptor.
> In terms of implementation, we would move relevant data from system.paxos on A to system.paxos on D.
> To implement this we would need to be very careful so that A' doesn't run concurrently on both A and D. So it would have to proceed in steps: we first block A from modifying A' state, then we migrate A' to D, then we allow D to modify A' state.
> This way, Synod invariants are preserved trivially: the set of acceptors was {A', B', C'}, and after bootstrapping it didn't change. So the allowed quorums are still the same: {A', B'}, {A', C'}, and {B', C'}. It's just that the acceptor A' is now running on a different node.

The problem with this approach is that it requires us to lock coordinators during migration, which will impact availability. Monotonic approaches which do this in two steps, first adding D then removing A, don't suffer from this. In https://github.com/scylladb/scylla/issues/5251 it was proposed to run `begin_and_repair_paxos()` before removing A to make sure that new quorums know everything established by old quorums.

**Comment by kbr-** [comment]
> What do you mean by stopping synod instances here?

I'm not completely sure if what I'm saying makes 100% sense, but lemme try to state it more precisely:

Before any node starts using the new tokens exclusively, say at time `T`, there must be a time `T' < T` such that the following must be true for any proposal `P` chosen after `T'`:
- the most recent commit used by `P` (i.e. the result of the previous instance) has been accepted and committed by a majority of acceptors *in the new set of replicas*.

In other words we must make sure that all "previous" instances (used by `P` to perform the CAS) have "finished" (were accepted and committed) on the new set of replicas too.

This might require refreshing all proposals at some point during the streaming phase using the new set of acceptors (additionally to the old set of acceptors), even if no queries are being made for the corresponding partitions. Or we could simulate that by copying data from system.paxos tables in old acceptors to new acceptors.

**Comment by kbr-** [comment]
Below is an algorithm to satisfy the property stated above.
I don't know if the steps stated below are sufficient for correctness, but whatever algorithm we end up with, it probably will have to be perform at least the steps described below (they are probably necessary).

1. make sure that every new proposal requires promises from a majority of new replicas (additionally to a majority of old replicas) before it's issued. This requires contacting every coordinator.
2. make sure that all proposals that don't use new replicas are either finished or abandoned. This requires contacting every acceptor and telling them not to accept any more of these "old" proposals (e.g. by making sure that all proposals accepted from now on have ballots higher than some threshold; i.e., the acceptors make a promise for this threshold; after step 1., every coordinator issues proposals with higher ballots).
3. refresh all previous proposals so that they are accepted by a majority of new replicas. This requires iterating through system.paxos tables on all old replicas. For every (row_key, cf_id) pair, we need to:
- choose a ballot,
- get a promise from a majority in old and a majority in new replicas not to accept any lower ballot,
- calculate the highest-numbered accepted proposal among the old replicas, AND new replicas (due to step 1., proposals might've been accepted on new replicas too), among proposals lower than our ballot,
- take the value of this highest-numbered proposal,
- make a proposal with our ballot and value, and
- get it accepted among a majority of old and new replicas.
4. get the proposal from step 3. committed on a majority of old and new replicas.
5. now we know all further committed proposals require new replicas to accept and commit too (due to step 1), and all proposals that were "in-flight" which used only the old replicas were refreshed to use the new replicas too. So we can allow coordinators to switch to using only the new replicas.

Step 3. seems like a lot of work. We need to do something that is equivalent to step 3., not necessarily do the standard promise-accept-learn cycle, but e.g. copying data from old to new acceptors (but we need to be careful to do this right; for example, we need to make sure that the value from the highest-numbered accepted proposal is copied to the new acceptors).
Also, I'm not sure that step 4. is strictly necessary.

**Comment by kbr-scylla** [comment]
> The fix suggested by Gleb is to have the topology change coordinator synchronize with each node before it starts streaming by issuing a call which makes sure the node sees the new topology and waits for all requests started earlier (which may use the old topology). Requests would be tracked using phased_barrier. This makes correctness insensitive to the timing of events.

We partially do this, except we allow drain to fail, in which case we advance fence version. But fencing is not implemented for Paxos yet. So I guess fencing for Paxos is the only missing ingredient to fix this?
